### PR TITLE
Support config via environment variable

### LINKED ISSUE #1440: Support config via environment variable
Our current model for configurations is:

CLI Arguments > Config File > Hard-coded Defaults

However, another common way to configure applications is via environment variables. They usually sit between the config file and the CLI arguments:

CLI Arguments > Environment Variables > Config File > Hard-coded Defaults

Because they sit between CLI and the contents of the config file, user cannot easily implement this on their end.

### PR DESCRIPTION
**Pull request by Neverlord**
Closes #1440.

I've split the tests for `config_option`. The parts that made sense to keep in unit tests moved to a new test suite with the new testing framework. The high-level tests are now implemented in Robot tests with baseline checks. I find that much easier to work with and found a bug in the dump-content output when nesting dictionaries in lists.

### CODE REVIEW

**Review by riemass — CHANGES_REQUESTED** [review]
Looking forward to this. Got a few questions and some change requests first.

File: libcaf_core/caf/config_value_writer.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ config_value_writer::~config_value_writer() {
 
 // -- interface functions ------------------------------------------------------
 
-bool config_value_writer::begin_object(type_id_t type, std::string_view) {
+bool config_value_writer::begin_object(type_id_t, std::string_view) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/config_value_writer.cpp:45]: Yes, all inspectors must have offer this signature. Whether they care about the type or now.

File: libcaf_core/caf/config_value_writer.cpp [hunk 2]
```diff
@@ -83,8 +83,6 @@ bool config_value_writer::begin_object(type_id_t type, std::string_view) {
     });
   if (!visit(f, st_.top()))
     return false;
-  if (type != invalid_type_id)
-    put(*get<settings*>(st_.top()), "@type", query_type_name(type));
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/config_value_writer.cpp:87]: Inserting the `@type` annotation here does nothing - except bloating the output `--dump-content`. CAF needs this annotation only if there a several options for a type. This is the case for variant fields. And in that case, we do still add the `@type` annotation. I suspect this annotation here was simply a leftover from early experiments, so I've dropped it.

File: libcaf_core/caf/config_option.cpp [hunk 3]
```diff
@@ -17,6 +17,27 @@ using std::string;
 
 namespace caf {
 
+namespace {
+
+template <class OutputIterator>
+auto copy_uppercase(std::string_view str, OutputIterator dst) {
+  for (auto c : str) {
+    if (isalnum(c))
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/config_option.cpp:25]: Everything that returns false for `isalnum` will be mapped to `_`. We could technically do separate checks for numbers and characters to only pass `isalpha`-characters to `toupper`, but `toupper` does that check anyways and the logic is more compact this way.

File: robot/config/options-driver.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+struct out_t {};
+
+out_t& operator<<(out_t& o, const std::string& str) {
+  puts(str.c_str());
+  return o;
+}
+
+int caf_main(actor_system&, const config& cfg) {
+  out_t out{};
+  out << "-- member variables --"s
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply robot/config/options-driver.cpp:71]: Just a little helper to keep the code more concise. If we'd use `std::cout`, we'd have to add a `<< '\n'` after each string, `out_t` will add a newline for us automatically.

File: robot/config/options.robot [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+*** Variables ***
+${BINARY_PATH}          /path/to/the/test/binary
+${CONFIG_FILE_DIR}      /path/to/the/config/file/dir
+
+*** Test Cases ***
+Test Files
+    FOR  ${i}  IN RANGE  1  5
+      Run Configuration File  ${CONFIG_FILE_DIR}/options-test${i}.cfg
+    END
+
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply robot/config/options.robot:16]: Hm... that raises a good point. Not everything should be configurable using the environment variables. Currently, you can set `HELP=true` to get the `--help` output. I'll fix that.

Regarding the `config-file` parameter, I'm a bit torn on whether we want to support that via environment variable.

File: libcaf_core/caf/config_option.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (riemass) [reply libcaf_core/caf/config_option.cpp:25]: It's fine as is, I've missed the mapping to `_`.

File: robot/config/options.robot (hunk 5 above)
[the remark this answers is not in the document]
Comment (Neverlord) [reply robot/config/options.robot:16]: Well, I also couldn't find a strong reason to not support it so... Done and tested.

[the remark this answers is not in the document]
Comment (riemass) [reply robot/config/options.robot:16]: Fully acknowledge, I didn't think of `help`, but that one was too volatile to be left.

**Review by riemass — APPROVED** [review]
Quick and efficient. Glad to approve this :+1:

**Pull request merged**